### ISSUE TITLE
building on windows

### ISSUE THREAD

**Issue by RyanHope** [fraimondo/cudaica#2]
At one point I thought you had a windows binary available but maybe I am mistaken. Anyway, I am all setup with a msys2/mingw64 build environment and CUDA and visual studio 12. How do I get this to compile? I was able to hobble my way through the configure but then it still fails on the build because of sys/mman.h missing and unistd.h.

**Comment by fraimondo** [comment]
I've never even tried building it on windows.

It can be ported. I guess CUDA behaves the same, it's just a matter of low level IO functions.

**Comment by RyanHope** [comment]
Yeah I bet it is just a few lines of code. Let me know if there is anything I can do to help out... if you need any more detailed build logs.

**Comment by chrisblais** [comment]
Super interested in a version that runs on windows.

**Comment by RyanHope** [comment]
Any update on a windows build attempt?

**Comment by fraimondo** [comment]
Not really.

I don’t have time right now. I don’t see how this could happen soon, at least from my side.

If there is some other developer that is willing to do it, I can review the code. But not much more.

**Comment by yhz-1995** [comment]
I am trying to modify the code and try to run it on Windows using Visual Studio. I think I have managed to deal with the low level IO functions, but stuck at the cuBLAS and CBLAS part. Visual Studio keeps telling me that I have unresolved external symbol issues, including: "dcopy_", "dgemm_", "dgesv_", "dgetrf_", "dgetri_", "dscal_", "dswap_", "dsyev_", "dsymm_", "dsyrk", "cublasCreat_v2", "cublasDaxpy_v2", "cublasDcopy_v2", "cublasDestroy_v2", "cublasDgemm_v2", "cublasDsyrk_v2". 

I have included "cblas.h" and "cublas_v2.h", but I guess cublas and cblas is not installed on Windows, so that these functions are not defined. I may try to install on Windows, or do somebody else have suggestions to workaround this problem?

**Comment by CloudyDory** [comment]
I have successfully build CUDAICA on Windows 10 using Visual Studio 2015, and it seems running fast and well now. 

@fraimondo If you have time, could you please help to review the code a little bit? Both the low-level IO part and the CUDA and BLAS part have issues, and I used some workarounds to bypass them. If you agree, I will upload the code on GitHub.

Thanks,

**Comment by fraimondo** [comment]
Hi Yunhui,

That’s great.

Do you have the code on GitHub so I can take a look?

**Comment by CloudyDory** [comment]
Hello Fraimondo,

I have uploaded modified version of CUDAICA on <https://github.com/yhz-1995/cudaica_win>.

I find that the cublas part of "step3" in infomax.cu is very slow, On my machine (i7 7700HQ + Nvidia 1050Ti) a lot of time is spent on transform the triangular matrix "u*u'" into a dense matrix. Instead, your "step3<<<>>>" kernel is very fast, but somehow you disabled it. Once I reenable this kernel, I get a 5x speed up on my machine.
